### ISSUE TITLE
FakeTensor RuntimeError when running fx.Interpreter in fake_mode while "dynamic" is set to True

### ISSUE THREAD

**Issue by kbadz-pl** [pytorch/pytorch#97250]
### 🐛 Describe the bug

When "dynamic" parameter to torch.compile is set to True, it is not possible to do fx.Interpreter -based propagation as just running the interpreter will cause Runtime Error.
![image](https://user-images.githubusercontent.com/122799060/226637281-9a69ca21-8205-4565-ac79-6c337c909e1e.png)
Attaching repro file: 
[repro.txt](https://github.com/pytorch/pytorch/files/11029369/repro.txt)

This error is not visible when running with dynamic=False.
Issue is visible on both 2.0 RC and latest 2.1.

### Error logs

![image](https://user-images.githubusercontent.com/122799060/226638614-bcf54ac2-665b-45d9-9abf-f8d48dde7280.png)

### Versions

Collecting environment information...
PyTorch version: 2.1.0.dev20230321+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: 10.0.0-4ubuntu1 
CMake version: version 3.25.0
Libc version: glibc-2.31

Python version: 3.9.16 (main, Dec  7 2022, 01:11:51)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.10.147+-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 525.85.12
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.7.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 48 bits virtual
CPU(s):                          2
On-line CPU(s) list:             0,1
Thread(s) per core:              2
Core(s) per socket:              1
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           85
Model name:                      Intel(R) Xeon(R) CPU @ 2.00GHz
Stepping:                        3
CPU MHz:                         2000.180
BogoMIPS:                        4000.36
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       32 KiB
L1i cache:                       32 KiB
L2 cache:                        1 MiB
L3 cache:                        38.5 MiB
NUMA node0 CPU(s):               0,1
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Mitigation; PTE Inversion
Vulnerability Mds:               Vulnerable; SMT Host state unknown
Vulnerability Meltdown:          Vulnerable
Vulnerability Mmio stale data:   Vulnerable
Vulnerability Retbleed:          Vulnerable
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1:        Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:        Vulnerable, IBPB: disabled, STIBP: disabled, PBRSB-eIBRS: Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Vulnerable
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat md_clear arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.24.1
[pip3] pytorch-triton==2.1.0+e650d3708b
[pip3] torch==2.1.0.dev20230321+cu117
[pip3] torchaudio==2.1.0.dev20230321+cu117
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.14.1
[pip3] torchvision==0.16.0.dev20230321+cu117
[conda] Could not collect

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh

**Comment by kbadz-pl** [comment]
So am I now guaranteed to be in fake mode when using AOTA? I use this compiler for FWD, BWD and optimizer and I remember there were some corner cases. If I’m not guaranteed - is there a way to check whether I am already in fake mode to not double it?

**Comment by SherlockNoMad** [comment]
AFAIK, we don't have a util function to check if we under FakeTensorMode, but it should relative simple to implement one. 
See _get_current_dispatch_mode_stack() in torch/utils/_python_dispatch.py 
You can walk through the mode stack to check if there is any FakeTensorMode. 

cc @ezyang and @eellison

**Comment by kbadz-pl** [comment]
FYI - so I checked it and it seems it does not have fake_mode nor fake_inputs for optimizers. Code that makes it working looks like this:
![image](https://user-images.githubusercontent.com/122799060/226862055-e4864584-6387-4085-bb2e-6107cc113e4c.png)

Is this what is expected? Or do you plan to make optimizer working same way like FWD/BWD? @SherlockNoMad

**Comment by ezyang** [comment]
It seems to me more like we shouldn't have applied fake tensor mode before calling the compiler. Let me see if this is an easy fix.

**Comment by ezyang** [comment]
What exactly is your optimizers code? Reading aot autograd, optimizers would be the "inference" case, which still should get fake mode since everything goes through the dispatcher.

**Comment by kbadz-pl** [comment]
@ezyang so I rechecked it from scratch. It looks like it's not optimizer but BWD. Here is script file I used for repro with the optimizer and bwd: 
[repro_dynamic.txt](https://github.com/pytorch/pytorch/files/11118847/repro_dynamic.txt)

Output result is this:
![image](https://user-images.githubusercontent.com/122799060/229042219-45f1b6e5-7985-45b6-8504-800d70565511.png)


I use SGD there, because when I try Adam then it just crashes on assertion error inside dynamo:
![image](https://user-images.githubusercontent.com/122799060/229042024-c2856a5c-d7d9-4261-b9d3-779b42e92e17.png)


Above was tested on 2.0, because running latest 2.1 results in even stranger crashes (fake_tensor issues between the BWD and OPT):
![image](https://user-images.githubusercontent.com/122799060/229042967-9e908528-55ee-4d7b-95d1-b281d16851dd.png)


Anyway, in relation to this specific ticket the issue is that compiler does not always get fake_mode and fake example_inputs, in this case it's BWD.

**Comment by ezyang** [comment]
OK, while I am not exactly sure what the problem is, I do know that dynamic=True backwards is going to get rewritten on master very shortly, which may help. I'll make sure to try your repro when this happens

**Comment by ezyang** [comment]
check if https://github.com/pytorch/pytorch/pull/99089 solves your problem

**Comment by kbadz-pl** [comment]
I took the latest PT2.1 and changed my hacky fakification with what you're using now (detect_fake_mode/FakeTensorProp way). Both above examples are working, Adam optimizer is also working fine now. Thanks! 

I also saw that you are planning to gradually remove the fake_mode completely from upper parts of the stack but this code should detect it and assign default FakeMode to it in such case so I don't expect additional issues here (hopefully).
